File: mcp/client/examples/stdio_example.py
"""
Example: Using MCP Client with Stdio Transport

This example demonstrates connecting to a local MCP server
using the stdio transport (spawns a process).
"""

import asyncio
from mcp.client import (
    create_mcp_client,
    MCPClientConfig,
    MCPTransportConfig,
    MCPTransportType,
    MCPToolUtils,
)


async def main():
    # Create client with stdio transport
    config = MCPClientConfig(
        name="stdio-example",
        version="1.0.0",
        transport=MCPTransportConfig(
            type=MCPTransportType.STDIO,
            command="npx",
            args=["-y", "@modelcontextprotocol/server-filesystem", "/tmp"]
        ),
        timeout=30000,
        debug=True,
    )

    client = create_mcp_client(config)

    # Set up event listeners
    def on_connected():
        print("✓ Connected to MCP server")

    def on_disconnected():
        print("✓ Disconnected from MCP server")

    def on_error(error):
        print(f"✗ Error: {error}")

    def on_state_change(state):
        print(f"State changed to: {state}")

    client.on("connected", on_connected)
    client.on("disconnected", on_disconnected)
    client.on("error", on_error)
    client.on("state_change", on_state_change)

    try:
        # Connect to server
        print("Connecting to MCP server...")
        server_info = await client.connect()

        print("\nServer Information:")
        print(f"  Name: {server_info.name}")
        print(f"  Version: {server_info.version}")
        print(f"  Protocol: {server_info.protocol_version}")
        print(f"  Capabilities: {server_info.capabilities}")

        # List available tools
        print("\nListing available tools...")
        tools = await client.tools.list_tools()

        print(f"\nFound {len(tools)} tools:")
        for index, tool in enumerate(tools, 1):
            print(f"\n{index}. {tool.name}")
            if tool.description:
                print(f"   Description: {tool.description}")
            print(f"   Input Schema: {tool.input_schema}")

        # Example: Call a tool if available
        if tools:
            first_tool = tools[0]
            print(f"\nCalling tool: {first_tool.name}")

            try:
                from mcp.client import MCPToolCall

                result = await client.tools.call_tool(
                    MCPToolCall(
                        name=first_tool.name,
                        arguments={}  # Adjust based on actual tool requirements
                    )
                )

                print("\nTool Result:")
                formatted = MCPToolUtils.format_result(result)
                print(formatted)

                if MCPToolUtils.has_error(result):
                    print("⚠️  Tool execution returned an error")

            except Exception as e:
                print(f"✗ Failed to call tool: {e}")

        # List resources
        print("\nListing available resources...")
        resources = await client.list_resources()

        print(f"\nFound {len(resources)} resources:")
        for index, resource in enumerate(resources, 1):
            print(f"\n{index}. {resource.name}")
            print(f"   URI: {resource.uri}")
            if resource.description:
                print(f"   Description: {resource.description}")
            if resource.mime_type:
                print(f"   MIME Type: {resource.mime_type}")

    except Exception as e:
        print(f"✗ Error: {e}")

    finally:
        # Disconnect
        print("\nDisconnecting...")
        await client.disconnect()
        print("Done!")


if __name__ == "__main__":
    asyncio.run(main())

File: packages/mcp-client/examples/stdio-example.ts
/**
 * Example: Using MCP Client with Stdio Transport
 *
 * This example demonstrates connecting to a local MCP server
 * using the stdio transport (spawns a process).
 */

import { createMCPClient, MCPTransportType, MCPToolUtils } from '../src';

async function main() {
  // Create client with stdio transport
  const client = createMCPClient({
    name: 'stdio-example',
    version: '1.0.0',
    transport: {
      type: 'stdio' as MCPTransportType,
      command: 'npx',
      args: ['-y', '@modelcontextprotocol/server-filesystem', '/tmp'],
    },
    timeout: 30000,
    debug: true,
  });

  // Set up event listeners
  client.on('connected', () => {
    console.log('✓ Connected to MCP server');
  });

  client.on('disconnected', () => {
    console.log('✓ Disconnected from MCP server');
  });

  client.on('error', (error) => {
    console.error('✗ Error:', error.message);
  });

  client.on('stateChange', (state) => {
    console.log(`State changed to: ${state}`);
  });

  try {
    // Connect to server
    console.log('Connecting to MCP server...');
    const serverInfo = await client.connect();

    console.log('\nServer Information:');
    console.log(`  Name: ${serverInfo.name}`);
    console.log(`  Version: ${serverInfo.version}`);
    console.log(`  Protocol: ${serverInfo.protocolVersion}`);
    console.log(`  Capabilities:`, serverInfo.capabilities);

    // List available tools
    console.log('\nListing available tools...');
    const tools = await client.tools.listTools();

    console.log(`\nFound ${tools.length} tools:`);
    tools.forEach((tool, index) => {
      console.log(`\n${index + 1}. ${tool.name}`);
      if (tool.description) {
        console.log(`   Description: ${tool.description}`);
      }
      console.log(`   Input Schema:`, JSON.stringify(tool.inputSchema, null, 2));
    });

    // Example: Call a tool if available
    if (tools.length > 0) {
      const firstTool = tools[0];
      console.log(`\nCalling tool: ${firstTool.name}`);

      try {
        const result = await client.tools.callTool({
          name: firstTool.name,
          arguments: {}, // Adjust based on actual tool requirements
        });

        console.log('\nTool Result:');
        const formatted = MCPToolUtils.formatResult(result);
        console.log(formatted);

        if (MCPToolUtils.hasError(result)) {
          console.log('⚠️  Tool execution returned an error');
        }
      } catch (error) {
        console.error('✗ Failed to call tool:', error);
      }
    }

    // List resources
    console.log('\nListing available resources...');
    const resources = await client.listResources();

    console.log(`\nFound ${resources.length} resources:`);
    resources.forEach((resource, index) => {
      console.log(`\n${index + 1}. ${resource.name}`);
      console.log(`   URI: ${resource.uri}`);
      if (resource.description) {
        console.log(`   Description: ${resource.description}`);
      }
      if (resource.mimeType) {
        console.log(`   MIME Type: ${resource.mimeType}`);
      }
    });

  } catch (error) {
    console.error('✗ Error:', error);
  } finally {
    // Disconnect
    console.log('\nDisconnecting...');
    await client.disconnect();
    console.log('Done!');
  }
}

// Run the example
main().catch(console.error);
